docs/libvirt-node-networking.md

```diff
@@ -0,0 +1,263 @@
+interface, bound to the same host interfaces like all other `macvlan` devices.
+We can then route all host traffic through that `macvlan` device, to allow host
+to guest and guest to host communication ([1], [2]).
+
+#### Prepare the host network
+
+`virt-handler` would on startup do the following setup:
+
+First, assuming that `eth0` was detected as the main node interface, CNI will
+be invoked with the follwoing parameters:
+```json
+{
+    "name": "myvm-nodenetwork",
+    "type": "macvlan",
+    "master": "eth0",
+    "mode": "bridge",
+    "ipam": {
+    		"type": "dhcp"
+    	}
+}
+```
+that CNI plugin, which can be taken from the CNI example plugins, roughly does
+
+```bash
+DEV=eth0
+DEV_IP=$(ip addr show $DEV | grep -Po 'inet \K[\d.]+')
+DEV_SUBNET=$(echo "$DEV_IP" | cut -d"." -f1-3)
+
+# Create a host ipvlan through which the host and the guests can talk to each other
+ip link add mv4 link $DEV type macvlan mode bridge
+ip link set dev mv4 up
+
+# here we multiple options:
+# * use dhcp, it will do the default routes for us
+# * do dhcp without setting the gateway, we then have to set it manually
+# * unconfigure $DEV and set the IP on mv4 and manully change the default route
+# using an extra IP via dhcp sounds like the best solution.
+# We can leave all other routes which belong to $DEV untouched.
+# No matter what we do, we don't want any routes modified.
+ip addr add $DEV_SUBNET.13/24 dev mv4 noprefixroute
+
+# normally we already have the netork namespace of libvirt given.
+ip netns add ns1
+
+# create the macvlan link attaching it to the parent host $DEV
+ip link add mv1 link $DEV type macvlan mode bridge
+
+# move the new interface mv1 to the new namespace
+ip link set mv1 netns ns1
+
+# bring the interface up
+# ns1 will later be used by libvirt to attach macvtaps to
+ip netns exec ns1 ip link set dev mv1 up
+
+
+```
+
+After that invocation, we have a host macvlan, with an IP address assigned. The
+CNI dhcp plugin comes with a minimalistic dhcp client, which we can use to do
+the DHCP renewals. An important side note here is, that the minimalistic daemon
+returns the acquired IP alog side with routing rules suggested by the DHCP
+server, but they are not applied, so the host is not modified.
+
+#### Prepare the VM device
+
+When a VM contains a network configuration like this:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: HostNetwork
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+```
+For every `HostNetwork`, `virt-handler` let's libvirt create a new macvtap
+attached to the one prepared macvlan in the initial setup step in the libvirt
+namespace.  Ee will use the CNI `dhcp` IPAM plugin with a provided or generated
+MAC address. Based on this MAC address and the IPAM plugin invocation we end up
+with the correct IP and MAC pair. The MAC address will then be used by libvirt
```

Comment:
If we connect the VM to the controll network, then we do not want to retrieve the IP form CNI, but rather from any IP authority on the control network, or provided it ourselfs.
I'd imagine that we'd actually ignore any IP provided by IPAM.

Reply:
In the macvlan/non-bridge scenario we would need the IP address, to make sure that our nested k8s node can talk to the bare metal k8s node. With the IP we can add the required route on the VM start, and remove the route on the VM stop.